### PR TITLE
build.d: Add cxx-unittest

### PR DESCRIPTION
**Pull request by MoonlightSentinel**
This add the `cxx-unittest` target and all required variables from `posix.mak` to `build.d`.

They are restricted to `version(Posix)` because it would require additional work to handle e.g. Cygwin on Azure)

TODO:
- [ ] Delete now obsolete CXX and CXXFLAGS detection from `posix.mak` (?)

### CODE REVIEW

**Comment by MoonlightSentinel** [comment]
CC @marler8997 Just in case you've started working on this as well

**Comment by marler8997** [comment]
Have not done this one yet.  Thanks for doing it.

**Review by wilzbach — COMMENTED** [review]
Good work! Please don't get discouraged by my comments.

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: Yeah, those were just notes to keep track of the differences between Makefiles and `build.d` (and maybe aid in reviewing them).
I will remove them

File: src/build.d
Comment (ibuclaw) [reply src/build.d]: At minimum, these warnings/options are required.
```
-fno-exceptions -fno-rtti -fasynchronous-unwind-tables  -Wextra -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wno-long-long -Wno-variadic-macros -Wno-overlength-strings
```

File: src/build.d
Comment (ibuclaw) [inline src/build.d]: You can get rid of `-D__pascal=`, `-DTARGET_XXX=1`, and `-DDM_TARGET_CPU_XXX=1` defines.  Just keep `-DMARS=1`

File: src/build.d
Comment (ibuclaw) [inline src/build.d]: Some of these aren't C++ options (i.e: `-unittest`, `-cov`).  Probably you also want to remove any environment flag variable that doesn't have a D equivalent as well.

File: src/build.d
Comment (ibuclaw) [reply src/build.d]: Indeed, the `-g` flag is missing from the C++ build recipes, which can prove to be a minor annoyance when debugging the cxx-unittest program.

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: This is a relict from `posix.mak` but compilerInfo seems like a cleaner approach.

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: See below

File: src/build.d
Comment (wilzbach) [reply src/build.d]: FYI: `compilerInfo` is only available for `dmd`. It's not supported by LDC nor GDC yet.

Comment (wilzbach) [reply src/build.d]: See also: https://dlang.org/changelog/2.079.0.html#json_includes

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: No, the compiler invocation detects the precise version (e.g. 2.079) which is used in
https://github.com/dlang/dmd/blob/9e561f48e869a13e4435bc4aeab6420fa85a79ba/src/dmd/globals.h#L313
and
https://github.com/dlang/dmd/blob/9e561f48e869a13e4435bc4aeab6420fa85a79ba/src/dmd/root/rmem.h#L19

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: Then i would suggest to implement the JSON handling for DMD and keepthe current approach for ldc2/gdc

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: Suggestions? This variable expands to:
- frontendHeaders
- frontend
- glue
- backendHeaders
which contains all `intermediate` code (after lexing, before codegen)

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: Refer to my earlier comment about `hostDmdVernum`

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: @kinke Any differences regarding LDC?

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: Done

File: src/build.d
Comment (kinke) [reply src/build.d]: There are surely differences wrt. used switches, and those obviously also vary across compilers (gcc, clang, MSVC, ...). I can't be bothered to look them up now; additionally, we specifically reduce warnings verbosity for the DMD headers IIRC.

File: src/build.d
Comment (marler8997) [reply src/build.d]: You're only using this variable in one place...why have a variable at all?

File: src/build.d
Comment (marler8997) [inline src/build.d]: for consistency with everything else, use `.chain(<args>).array`

File: src/build.d
Comment (marler8997) [inline src/build.d]: `cxxKind = "-std=gnu++98"`

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: This variable is used by multiple targets in the Makefile which will also move to `build.d`

Comment (marler8997) [reply src/build.d]: I see.   Ok well I agree that we should use a better name than `sources`.  Maybe `cxxFrontendSources`?

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: ?

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: Can do, but that makes it much more verbose:
```suggestion
- sources = srcDir.buildPath("tests", "cxxfrontend.c") ~ .sources.sources ~ .sources.root;

+ sources = [srcDir.buildPath("tests", "cxxfrontend.c")].chain(.sources.sources, .sources.root).array;
or
+ sources = chain([srcDir.buildPath("tests", "cxxfrontend.c")], .sources.sources, .sources.root).array;
```

Comment (MoonlightSentinel) [reply src/build.d]: Can do, but that makes it much more verbose:
```suggestion
- sources = srcDir.buildPath("tests", "cxxfrontend.c") ~ .sources.sources ~ .sources.root;

+ sources = [srcDir.buildPath("tests", "cxxfrontend.c")].chain(.sources.sources, .sources.root).array;
or
+ sources = chain([srcDir.buildPath("tests", "cxxfrontend.c")], .sources.sources, .sources.root).array;
```

Comment (MoonlightSentinel) [reply src/build.d]: Slightly off topic, but it would be nice to use something like c++'s std::path instead of strings (which overloads / ) to improve the readability of `sources`, `target`, ...

Comment (wilzbach) [reply src/build.d]: Like Vibe.d's `NativePath`?

https://github.com/dlang/dub/blob/master/source/dub/internal/vibecompat/inet/path.d#L31

Phobos is in feature-freeze and we can't rely on dub packages here. So the only option is make it part of `build.d` which is already quite big.

Comment (wilzbach) [reply src/build.d]: > for consistency with everything else, use .chain(<args>).array

IIRC `chain` was initially only used where there were ranges. Even if it is now used more prominently, I don't think that we should be pedantic about it. It's really just useful if you work with non-arrays.

**Review by wilzbach — COMMENTED** [review]
We're still talking about +219 lines to add a single C++ frontend test :/

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: Yeah, should this be part of this PR?

Comment (wilzbach) [reply src/build.d]: Entirely up to you. I just noticed it while looking at the diff.

Comment (MoonlightSentinel) [reply src/build.d]: Then i will do the cleanup in another PR to keep these changes as small as possible

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: Done. It's slightly overkill for now but will be useful when we add support for other compilers

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: Sorry, old habits die hard

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: Agreed, but we should keep ENABLE_SANITIZERS and maybe ENABLE_COVERAGE as these can be beneficial for testing

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: Done

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: According to the [Wiki](https://wiki.dlang.org/DMD_Source_Guide) the `frontend` refers to all steps before the codegen, including lexing and parsing. But the corresponding source files are not included in `sources`.
Maybe something along the lines of `semantic`?

File: src/build.d
Comment (marler8997) [reply src/build.d]: I believe that the expression `"g++"` will always evaluate to `true` so this ternary expression will always return `"-std=gnu++98"`.

Comment (MoonlightSentinel) [reply src/build.d]: No,e.g Linux_64_64 on the autotester run clang

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: Done

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: Done

File: src/build.d
Comment (marler8997) [reply src/build.d]: goodness I thought the `==` was an `=`.  ignore this comment

File: src/posix.mak
Comment (marler8997) [inline src/posix.mak]: Might just want to add `CXX=$(HOST_CXX)` to line 393 `RUN_BUILD = ...`

**Comment by marler8997** [comment]
> We're still talking about +219 lines to add a single C++ frontend test :/

The trouble here is that specifying a target in our `build.d` at minimum takes around 15 lines of code, which would have been around 2 lines of Make.   Plus, we aren't using any "build-specific" libraries so everything is implemented inside our specific build.

One way to make this smaller would be to put the build-library code into the "tools" repo.  This would also allow other repos to use it as well.

File: src/build.d
Comment (MoonlightSentinel) [reply src/build.d]: Thankfully `const` prevents such debugging nightmares.

File: src/posix.mak
Comment (MoonlightSentinel) [reply src/posix.mak]: Done

**Comment by MoonlightSentinel** [comment]
I converted `cxxFrontend` and `cxxUnittestExe` to local functions (as they are only needed by `runCxxUnittest`) and compacted the `command` arrays to reduce the diff. Let me know what you think.

+1 for splitting `build.d`

File: src/posix.mak
Comment (marler8997) [reply src/posix.mak]: Looks like you forgot to remove the CCX=$(HOST_CXX) here

Comment (MoonlightSentinel) [reply src/posix.mak]: Whoops

File: src/build.d
Comment (marler8997) [reply src/build.d]: looks good, will approve once we change `sources.sources` to something more specific.  I'm not too particular, just more specific than `sources`.

Comment (MoonlightSentinel) [reply src/build.d]: Removed `sources` as this set of files is more confusing than spelling it out

**Review by marler8997 — APPROVED** [review]

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
Auto-merge toggled on

**Comment by MoonlightSentinel** [comment]
@thewilsonator These WIP commits should've been squashed before merging ...

**Pull request merged**
